File: src/scripts/lint_trust_coherence.py
#!/usr/bin/env python3
"""Lint trust/safety coherence across the discovery manifest.

Phase 5.4 of the monorepo trust-and-safety layer. Walks the freshly
built `dist/discovery/discovery-manifest.json` and asserts three
invariants:

  1. Every pack whose ``trust_summary`` declares ``advisory`` or
     ``restricted`` artefacts ships at least one ``*safety-floor*``
     rule in the same pack.
  2. Every artefact with ``trust.human_review_required: true`` carries
     the ``_HRR_BANNER_MARKER`` in its compiled output under
     ``.agent-src/<logical>``.
  3. Every rule listed in ``router.json`` ``kernel[]`` declares
     ``trust.level: core`` (no escalation to advisory/restricted,
     no demotion to experimental).

Exits 0 clean, 1 on any violation. Stdlib + pyyaml. Cap: ≤ 200 LOC.
"""
from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path
from typing import Any

ROOT = Path(__file__).resolve().parents[2]
MANIFEST = ROOT / "dist" / "discovery" / "discovery-manifest.json"
ROUTER = ROOT / "dist" / "router.json"
COMPILED_SRC = ROOT / ".agent-src"

sys.path.insert(0, str(Path(__file__).resolve().parent))
from _lib.agent_src import strip_source_prefix  # noqa: E402

# Imported lazily inside _banner_marker() to keep the cap loose if condense.py
# grows additional top-level side effects.
_BANNER_MARKER = "<!-- agent-config:human-review-banner -->"

# Trust levels that demand a domain-safety floor in the same pack.
_FLAGGED_LEVELS = ("advisory", "restricted")
_SAFETY_FLOOR_FRAGMENT = "safety-floor"


def _load_manifest(path: Path) -> dict[str, Any]:
    if not path.exists():
        raise SystemExit(
            f"ERROR: manifest not found: {path}\n"
            "  Run `task build-discovery` first."
        )
    return json.loads(path.read_text("utf-8"))


def _load_kernel(path: Path) -> set[str]:
    if not path.exists():
        raise SystemExit(f"ERROR: router.json not found: {path}")
    data = json.loads(path.read_text("utf-8"))
    kernel = data.get("kernel") or []
    if not isinstance(kernel, list):
        raise SystemExit("ERROR: router.json `kernel` must be a list")
    return {str(name) for name in kernel}


def _check_pack_safety_floors(manifest: dict[str, Any]) -> list[str]:
    """Check 1: advisory/restricted packs ship a *safety-floor* rule."""
    errs: list[str] = []
    # Build pack -> [artefact path] index from the artefact list so we can
    # spot the safety-floor regardless of how trust_summary was computed.
    pack_paths: dict[str, list[str]] = {}
    for art in manifest.get("artefacts", []):
        for pack in art.get("packs", []) or []:
            pack_paths.setdefault(pack, []).append(art["path"])

    for pack in manifest.get("packs", []):
        summary = pack.get("trust_summary", {}) or {}
        flagged_total = sum(int(summary.get(lvl, 0)) for lvl in _FLAGGED_LEVELS)
        if flagged_total == 0:
            continue
        paths = pack_paths.get(pack["id"], [])
        has_floor = any(_SAFETY_FLOOR_FRAGMENT in p for p in paths)
        if not has_floor:
            counts = ", ".join(
                f"{lvl}={int(summary.get(lvl, 0))}" for lvl in _FLAGGED_LEVELS
            )
            errs.append(
                f"pack `{pack['id']}` declares flagged artefacts ({counts})"
                f" but ships no `*{_SAFETY_FLOOR_FRAGMENT}*` rule"
            )
    return errs


def _check_human_review_banners(
    manifest: dict[str, Any], compiled_src: Path
) -> list[str]:
    """Check 2: every human_review_required artefact has the banner."""
    errs: list[str] = []
    for art in manifest.get("artefacts", []):
        trust = art.get("trust", {}) or {}
        if not trust.get("human_review_required"):
            continue
        rel = art["path"]
        logical = strip_source_prefix(rel)
        if logical is None:
            errs.append(
                f"{rel}: human_review_required=true but path is not under"
                " any known source root"
            )
            continue
        compiled = compiled_src / logical
        if not compiled.exists():
            errs.append(
                f"{rel}: human_review_required=true but compiled output"
                f" missing at `{compiled.relative_to(ROOT)}`"
            )
            continue
        body = compiled.read_text("utf-8", errors="replace")
        if _BANNER_MARKER not in body:
            errs.append(
                f"{rel}: human_review_required=true but compiled output"
                f" `{compiled.relative_to(ROOT)}` is missing the HRR banner"
                f" (`{_BANNER_MARKER}`) — re-run `task condense`."
            )
    return errs


def _check_kernel_trust(
    manifest: dict[str, Any], kernel: set[str]
) -> list[str]:
    """Check 3: every kernel rule declares trust.level=core."""
    errs: list[str] = []
    # name -> artefact for category=rule entries. Manifest does not always
    # populate `name` for rules, so fall back to the logical filename stem.
    rule_by_name: dict[str, dict[str, Any]] = {}
    for art in manifest.get("artefacts", []):
        if art.get("category") != "rule":
            continue
        name = art.get("name")
        if not name:
            logical = strip_source_prefix(art.get("path", ""))
            if logical is None:
                continue
            stem = Path(logical).stem
            name = stem
        rule_by_name[name] = art

    for kname in sorted(kernel):
        art = rule_by_name.get(kname)
        if art is None:
            errs.append(
                f"kernel rule `{kname}` listed in router.json but no"
                " matching artefact in manifest"
            )
            continue
        level = (art.get("trust", {}) or {}).get("level")
        if level != "core":
            errs.append(
                f"kernel rule `{kname}` has trust.level=`{level}`"
                " — must be `core` (router.json kernel guarantees Iron-Law"
                " floor)"
            )
    return errs


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    parser.add_argument("--quiet", action="store_true")
    parser.add_argument(
        "--manifest", type=Path, default=MANIFEST, help="discovery manifest"
    )
    parser.add_argument(
        "--router", type=Path, default=ROUTER, help="router.json with kernel[]"
    )
    parser.add_argument(
        "--compiled-src",
        type=Path,
        default=COMPILED_SRC,
        help="compiled output root (.agent-src/)",
    )
    args = parser.parse_args(argv)

    manifest = _load_manifest(args.manifest)
    kernel = _load_kernel(args.router)

    errs: list[str] = []
    errs.extend(_check_pack_safety_floors(manifest))
    errs.extend(_check_human_review_banners(manifest, args.compiled_src))
    errs.extend(_check_kernel_trust(manifest, kernel))

    if errs:
        for e in errs:
            print(f"ERROR: {e}", file=sys.stderr)
        print(
            f"\n{len(errs)} trust-coherence violation(s) across"
            f" {len(manifest.get('packs', []))} pack(s) and"
            f" {len(manifest.get('artefacts', []))} artefact(s).",
            file=sys.stderr,
        )
        return 1

    if not args.quiet:
        print(
            "✅  lint-trust-coherence:"
            f" {len(manifest.get('packs', []))} pack(s),"
            f" {len(kernel)} kernel rule(s),"
            f" {sum(1 for a in manifest.get('artefacts', []) if (a.get('trust') or {}).get('human_review_required'))}"
            " HRR artefact(s) clean."
        )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/scripts/lint_trust_coherence.ts
#!/usr/bin/env tsx
/**
 * Lint trust/safety coherence across the discovery manifest.
 *
 * TypeScript twin of `src/scripts/lint_trust_coherence.py` (ADR-092,
 * Phase 4 / Wave 4b). Mirrors the Python CLI contract EXACTLY — `--quiet`
 * / `--manifest` / `--router` / `--compiled-src` flags, finding messages,
 * stdout/stderr split (errors + summary on stderr; success on stdout),
 * SystemExit-on-missing-file semantics, exit codes. No behaviour changes —
 * latent bugs replicated.
 *
 * Walks the discovery manifest and asserts three invariants:
 *   1. advisory/restricted packs ship a `*safety-floor*` rule;
 *   2. every human_review_required artefact carries the HRR banner marker
 *      in its compiled output under dist/agent-src/;
 *   3. every router.json kernel[] rule declares trust.level: core.
 *
 * Exit codes: 0 clean, 1 on any violation. SystemExit (code 1, stderr ERROR)
 * when the manifest or router.json is missing.
 *
 * Test seam: `_setConfigForTest` mirrors the Python tests' monkeypatch of
 * module-level ROOT / MANIFEST / ROUTER / COMPILED_SRC constants.
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';
import { strip_source_prefix } from './_lib/agent_src.js';

const _HERE = fileURLToPath(import.meta.url);
const _DEFAULT_ROOT = path.resolve(path.dirname(_HERE), '..', '..');

interface Config {
    ROOT: string;
    MANIFEST: string;
    ROUTER: string;
    COMPILED_SRC: string;
}

function _deriveConfig(root: string): Config {
    return {
        ROOT: root,
        MANIFEST: path.join(root, 'dist', 'discovery', 'discovery-manifest.json'),
        ROUTER: path.join(root, 'dist', 'router.json'),
        COMPILED_SRC: path.join(root, 'dist/agent-src'),
    };
}

const _config: Config = _deriveConfig(_DEFAULT_ROOT);

/** Test seam mirroring the Python tests' monkeypatch of module constants. */
function _setConfigForTest(overrides: Partial<Config>): void {
    Object.assign(_config, overrides);
}

const _BANNER_MARKER = '<!-- agent-config:human-review-banner -->';

const _FLAGGED_LEVELS = ['advisory', 'restricted'] as const;
const _SAFETY_FLOOR_FRAGMENT = 'safety-floor';

// SystemExit carrier — mirrors Python `raise SystemExit(msg)` (prints msg to
// stderr, exits 1) when surfaced at the top level.
class SystemExit extends Error {
    constructor(public readonly msg: string) {
        super(msg);
        this.name = 'SystemExit';
    }
}

function _exists(p: string): boolean {
    try {
        fs.statSync(p);
        return true;
    } catch {
        return false;
    }
}

/** `path.stem` — filename without the final suffix. */
function _stem(p: string): string {
    const base = path.basename(p);
    const dot = base.lastIndexOf('.');
    return dot > 0 ? base.slice(0, dot) : base;
}

/** POSIX relative path of `target` under `root` (str(Path.relative_to)). */
function _relToPosix(target: string, root: string): string {
    return path.relative(root, target).split(path.sep).join('/');
}

type JsonObject = Record<string, unknown>;

function _asArray(v: unknown): unknown[] {
    return Array.isArray(v) ? v : [];
}

function _asObject(v: unknown): JsonObject {
    return v !== null && typeof v === 'object' && !Array.isArray(v) ? (v as JsonObject) : {};
}

function _load_manifest(p: string): JsonObject {
    if (!_exists(p)) {
        throw new SystemExit(
            `ERROR: manifest not found: ${p}\n` + '  Run `task build-discovery` first.',
        );
    }
    return JSON.parse(fs.readFileSync(p, 'utf-8')) as JsonObject;
}

function _load_kernel(p: string): Set<string> {
    if (!_exists(p)) {
        throw new SystemExit(`ERROR: router.json not found: ${p}`);
    }
    const data = JSON.parse(fs.readFileSync(p, 'utf-8')) as JsonObject;
    const kernel = data['kernel'] ?? [];
    if (!Array.isArray(kernel)) {
        throw new SystemExit('ERROR: router.json `kernel` must be a list');
    }
    return new Set(kernel.map((name) => String(name)));
}

function _check_pack_safety_floors(manifest: JsonObject): string[] {
    const errs: string[] = [];
    const packPaths = new Map<string, string[]>();
    for (const artRaw of _asArray(manifest['artefacts'])) {
        const art = _asObject(artRaw);
        for (const pack of _asArray(art['packs'])) {
            const key = String(pack);
            if (!packPaths.has(key)) {
                packPaths.set(key, []);
            }
            packPaths.get(key)!.push(String(art['path']));
        }
    }

    for (const packRaw of _asArray(manifest['packs'])) {
        const pack = _asObject(packRaw);
        const summary = _asObject(pack['trust_summary']);
        const flaggedTotal = _FLAGGED_LEVELS.reduce(
            (acc, lvl) => acc + _toInt(summary[lvl] ?? 0),
            0,
        );
        if (flaggedTotal === 0) {
            continue;
        }
        const paths = packPaths.get(String(pack['id'])) ?? [];
        const hasFloor = paths.some((p) => p.includes(_SAFETY_FLOOR_FRAGMENT));
        if (!hasFloor) {
            const counts = _FLAGGED_LEVELS.map(
                (lvl) => `${lvl}=${_toInt(summary[lvl] ?? 0)}`,
            ).join(', ');
            errs.push(
                `pack \`${String(pack['id'])}\` declares flagged artefacts (${counts})` +
                    ` but ships no \`*${_SAFETY_FLOOR_FRAGMENT}*\` rule`,
            );
        }
    }
    return errs;
}

function _check_human_review_banners(manifest: JsonObject, compiledSrc: string): string[] {
    const errs: string[] = [];
    for (const artRaw of _asArray(manifest['artefacts'])) {
        const art = _asObject(artRaw);
        const trust = _asObject(art['trust']);
        if (!trust['human_review_required']) {
            continue;
        }
        const rel = String(art['path']);
        const logical = strip_source_prefix(rel);
        if (logical === null) {
            errs.push(
                `${rel}: human_review_required=true but path is not under` +
                    ' any known source root',
            );
            continue;
        }
        const compiled = path.join(compiledSrc, logical);
        if (!_exists(compiled)) {
            errs.push(
                `${rel}: human_review_required=true but compiled output` +
                    ` missing at \`${_relToPosix(compiled, _config.ROOT)}\``,
            );
            continue;
        }
        const body = fs.readFileSync(compiled, 'utf-8');
        if (!body.includes(_BANNER_MARKER)) {
            errs.push(
                `${rel}: human_review_required=true but compiled output` +
                    ` \`${_relToPosix(compiled, _config.ROOT)}\` is missing the HRR banner` +
                    ` (\`${_BANNER_MARKER}\`) — re-run \`task condense\`.`,
            );
        }
    }
    return errs;
}

function _check_kernel_trust(manifest: JsonObject, kernel: Set<string>): string[] {
    const errs: string[] = [];
    const ruleByName = new Map<string, JsonObject>();
    for (const artRaw of _asArray(manifest['artefacts'])) {
        const art = _asObject(artRaw);
        if (art['category'] !== 'rule') {
            continue;
        }
        let name = art['name'];
        if (!name) {
            const logical = strip_source_prefix(String(art['path'] ?? ''));
            if (logical === null) {
                continue;
            }
            name = _stem(logical);
        }
        ruleByName.set(String(name), art);
    }

    for (const kname of [...kernel].sort()) {
        const art = ruleByName.get(kname);
        if (art === undefined) {
            errs.push(
                `kernel rule \`${kname}\` listed in router.json but no` +
                    ' matching artefact in manifest',
            );
            continue;
        }
        const level = _asObject(art['trust'])['level'];
        if (level !== 'core') {
            errs.push(
                `kernel rule \`${kname}\` has trust.level=\`${_pyStr(level)}\`` +
                    ' — must be `core` (router.json kernel guarantees Iron-Law' +
                    ' floor)',
            );
        }
    }
    return errs;
}

/** Mirror Python `f"{level}"` for a possibly-None value. */
function _pyStr(v: unknown): string {
    if (v === null || v === undefined) {
        return 'None';
    }
    if (typeof v === 'boolean') {
        return v ? 'True' : 'False';
    }
    return String(v);
}

/** Mirror Python `int(x)` for the int-ish summary counts. */
function _toInt(v: unknown): number {
    if (typeof v === 'number') {
        return Math.trunc(v);
    }
    if (typeof v === 'boolean') {
        return v ? 1 : 0;
    }
    const n = parseInt(String(v), 10);
    return Number.isNaN(n) ? 0 : n;
}

interface ParsedArgs {
    quiet: boolean;
    manifest: string;
    router: string;
    compiledSrc: string;
}

function _argparse_error(message: string): never {
    process.stderr.write(`lint_trust_coherence: error: ${message}\n`);
    process.exit(2);
}

function parse_args(argv: readonly string[]): ParsedArgs {
    let quiet = false;
    let manifest = _config.MANIFEST;
    let router = _config.ROUTER;
    let compiledSrc = _config.COMPILED_SRC;
    for (let i = 0; i < argv.length; i++) {
        const arg = argv[i]!;
        const takeValue = (flag: string): string => {
            const v = argv[++i];
            if (v === undefined) {
                _argparse_error(`argument ${flag}: expected one argument`);
            }
            return v;
        };
        if (arg === '--quiet') {
            quiet = true;
        } else if (arg === '--manifest') {
            manifest = takeValue('--manifest');
        } else if (arg.startsWith('--manifest=')) {
            manifest = arg.slice('--manifest='.length);
        } else if (arg === '--router') {
            router = takeValue('--router');
        } else if (arg.startsWith('--router=')) {
            router = arg.slice('--router='.length);
        } else if (arg === '--compiled-src') {
            compiledSrc = takeValue('--compiled-src');
        } else if (arg.startsWith('--compiled-src=')) {
            compiledSrc = arg.slice('--compiled-src='.length);
        } else if (arg === '-h' || arg === '--help') {
            process.stdout.write(
                'usage: lint_trust_coherence [-h] [--quiet] [--manifest MANIFEST] ' +
                    '[--router ROUTER] [--compiled-src COMPILED_SRC]\n',
            );
            process.exit(0);
        } else {
            _argparse_error(`unrecognized arguments: ${arg}`);
        }
    }
    return { quiet, manifest, router, compiledSrc };
}

function main(argv: readonly string[] = []): number {
    const args = parse_args(argv);

    const manifest = _load_manifest(args.manifest);
    const kernel = _load_kernel(args.router);

    const errs: string[] = [];
    errs.push(..._check_pack_safety_floors(manifest));
    errs.push(..._check_human_review_banners(manifest, args.compiledSrc));
    errs.push(..._check_kernel_trust(manifest, kernel));

    const packCount = _asArray(manifest['packs']).length;
    const artCount = _asArray(manifest['artefacts']).length;

    if (errs.length > 0) {
        for (const e of errs) {
            process.stderr.write(`ERROR: ${e}\n`);
        }
        process.stderr.write(
            `\n${errs.length} trust-coherence violation(s) across` +
                ` ${packCount} pack(s) and` +
                ` ${artCount} artefact(s).\n`,
        );
        return 1;
    }

    if (!args.quiet) {
        const hrrCount = _asArray(manifest['artefacts']).filter(
            (a) => _asObject(_asObject(a)['trust'])['human_review_required'],
        ).length;
        process.stdout.write(
            '✅  lint-trust-coherence:' +
                ` ${packCount} pack(s),` +
                ` ${kernel.size} kernel rule(s),` +
                ` ${hrrCount}` +
                ' HRR artefact(s) clean.\n',
        );
    }
    return 0;
}

const _isCliEntry =
    process.argv[1] !== undefined &&
    import.meta.url === pathToFileURL(path.resolve(process.argv[1])).href;
if (_isCliEntry || process.argv[1] === _HERE) {
    try {
        process.exit(main(process.argv.slice(2)));
    } catch (e) {
        if (e instanceof SystemExit) {
            process.stderr.write(`${e.msg}\n`);
            process.exit(1);
        }
        throw e;
    }
}

export {
    type Config,
    _setConfigForTest,
    _config,
    _BANNER_MARKER,
    SystemExit,
    main,
};
